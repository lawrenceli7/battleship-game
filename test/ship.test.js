import { describe, expect, test } from "@jest/globals";
import Ship from "../src/js/ship";

describe("ship length", () => {
  test("has correct length", () => {
    expect(new Ship(4)).toHaveProperty("length", 4);
    expect(new Ship(2)).toHaveProperty("length", 2);

    expect(new Ship(4)).toMatchSnapshot();
    expect(new Ship(2)).toMatchSnapshot();
  });

  test("throws if length not between 2 and 5", () => {
    expect(() => {
      new Ship(0);
    }).toThrow();
    expect(() => {
      new Ship(1);
    }).toThrow();
    expect(() => {
      new Ship(2);
    }).not.toThrow();
    expect(() => {
      new Ship(3);
    }).not.toThrow();
    expect(() => {
      new Ship(4);
    }).not.toThrow();
    expect(() => {
      new Ship(5);
    }).not.toThrow();
    expect(() => {
      new Ship(6);
    }).toThrow();
    expect(() => {
      new Ship(7);
    }).toThrow();

    expect(new Ship(2)).toMatchSnapshot();
    expect(new Ship(4)).toMatchSnapshot();
  });
});

describe("receiving hits", () => {
  test("keeps track of number of times hit", () => {
    const ship = new Ship(4);
    expect(ship).toHaveProperty("hits", 0);

    ship.hit();
    ship.hit();
    ship.hit();
    expect(ship).toHaveProperty("hits", 3);

    ship.hit();
    expect(ship).toHaveProperty("hits", 4);

    expect(ship).toMatchSnapshot();
  });

  test("caps number of hits at ship length", () => {
    const ship = new Ship(3);
    ship.hit();
    ship.hit();
    ship.hit();
    ship.hit();
    expect(ship).toHaveProperty("hits", 3);

    expect(ship).toMatchSnapshot();
  });

  test("is sunk if number of hits equals ship length", () => {
    const ship = new Ship(2);
    expect(ship.isSunk()).toBe(false);
    ship.hit();
    expect(ship.isSunk()).toBe(false);
    ship.hit();
    expect(ship.isSunk()).toBe(true);

    expect(ship).toMatchSnapshot();
  });
});
