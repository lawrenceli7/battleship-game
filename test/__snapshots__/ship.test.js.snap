// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`receiving hits caps number of hits at ship length 1`] = `
Ship {
  "hits": 3,
  "length": 3,
}
`;

exports[`receiving hits is sunk if number of hits equals ship length 1`] = `
Ship {
  "hits": 2,
  "length": 2,
}
`;

exports[`receiving hits keeps track of number of times hit 1`] = `
Ship {
  "hits": 4,
  "length": 4,
}
`;

exports[`ship length has correct length 1`] = `
Ship {
  "hits": 0,
  "length": 4,
}
`;

exports[`ship length has correct length 2`] = `
Ship {
  "hits": 0,
  "length": 2,
}
`;

exports[`ship length throws if length not between 2 and 5 1`] = `
Ship {
  "hits": 0,
  "length": 2,
}
`;

exports[`ship length throws if length not between 2 and 5 2`] = `
Ship {
  "hits": 0,
  "length": 4,
}
`;
